### DISCUSSION TITLE
[UX] allow object creation while dragging cord?

### DISCUSSION THREAD

**Discussion by ben-wes** [pure-data/pure-data#2371] [Ideas]
similar to the creation of connected objects via shortcut from a selected object (ctrl+1 etc.), it could be an option to create objects while dragging a cord from an outlet.

currently, this creates a bit of a glitchy state and the connection disappears again when the object is instantiated ...

![image](https://github.com/pure-data/pure-data/assets/1923787/bd9309b5-be89-4146-91cc-498f0c89d793)

...  but it's already close to what i imagine anyway. :)

the object should probably still be dragged around with the cable until the mouse button is released in this case.

**Comment by umlaeute** [comment]
i wonder how useful that would actually be in practice.

**Reply by ben-wes** [reply]
i'm sure one would need to get used to this option first - but at least for the case of adding objects to 2nd or further right outlets, it would make things a bit faster. currently that requires:
1. create object, enter name
2. deselect or switch selection with tab (while thinking about it a bit more, maybe this here is the actual issue - and would be solved with the first suggestion of #857)
3. create connection from 2nd outlet above to inlet of new object
4. reselect to continue patching with keyboard (if the selection wasn't preserved)

with the new option, it would be:
1. pull cord from 2nd outlet above
2. create object, enter name

... so in a way, this might be obsolete if instantiating objects with preserved selection would be (more easily) possible.

**Reply by umlaeute** [reply]
i think the workflows you sketched are missing a few bits (I guess you just started in the middle after the 1st object was somehow in the state you imagined; i think it is helpful to not emit these steps)

the "old way" should read:

1. create 1st object (<kbd>Ctrl</kbd>+<kbd>1</kbd>), enter name
2. deselect, so that the object instantiates
3. create 2nd object (<kbd>Ctrl</kbd>+<kbd>1</kbd>), enter name
4. deselect or re-select (or switch selection with <kbd>Tab</kbd>) 2nd object (so it instantiates)
5. create connection from right-hand outlet of 1st object to inlet of 2nd object
6. (optionally) re-select 2nd object to continue patching with keyboard (if the selection wasn't preserved)
7. ...


the "new way" seems to be an oversimplification (for the sake of the argument) from what is actually necessary.
this seems to be more like it:
1. create 1st object (<kbd>Ctrl</kbd>+<kbd>1</kbd>), enter name
2. deselect (or switch selection with <kbd>Tab</kbd>), so that the object instantiates
3. pull cord from right-hand outlet of newly created object
4. create 2nd object (<kbd>Ctrl</kbd>+<kbd>1</kbd>), enter name
5. ...

my point is that the you need to explicitly instantiate new objects, for which there are various ways (as you have rightfully pointed out this might well be the root issue here).
instantiating takes quite a bit of user interaction, as it either requires a mouse-click, or switches the selection as a side-effect (e.g. when using <kbd>Tab</kbd>). in some cases (namely: "old way \#2") we must make sure no object is selected at all (which could be done with calling "Select All" twice - <kbd>Ctrl</kbd>+<kbd>a</kbd>-<kbd>Ctrl</kbd>+<kbd>a</kbd> - though I've never did that before).
i don't see (yet) how the 1st suggestion in #857 is actually going to help with this last case.

anyhow: the new way would only require a single de-selection of a newly created object (and we could even use <kbd>Tab</kbd> for that, as we don't care about what is selected afterwards), so it would definitely cut down on the biggest issue.

**Reply by ben-wes** [reply]
> we must make sure no object is selected at all

took me a bit to see what you meant here - and sure: currently, in order to create the second object at a freely chosen position, everything must be deselected before. so "old way, 2" is probably easiest with <kbd>Esc</kbd>. and #857 is not related there at all.

the option to instantiate with preserved selection is only relevant for "old way, 4", since you could continue with keyboard-patching after step 5 then.

**Reply by umlaeute** [reply]
ah yes, i forgot about <kbd>Esc</kbd> :-)
